### PR TITLE
A sample application showing how to connect Google Home to .NET Conta…

### PR DESCRIPTION
**Pull request by meteatamel**
…iners running on Google Cloud

### CODE REVIEW

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+            if (errorMessage != null) return DialogflowApp.Tell(errorMessage);
+
+            var bigQueryClient = await BigQueryClient.CreateAsync(Program.AppSettings.GoogleCloudSettings.ProjectId);
+
+            // Build the parameterized SQL query
+            var tableGsod = bigQueryClient.GetTable("bigquery-public-data", "noaa_gsod", $"gsod*");
+            var tableStations = bigQueryClient.GetTable("bigquery-public-data", "noaa_gsod", "stations");
+            var sql = $"SELECT (gsod.temp - 32)*5/8 AS celcius, stations.name AS name, gsod.year AS y, gsod.mo AS m, gsod.da AS d \n" +
+                $"FROM {tableGsod} AS gsod \n" +
```
Comment (SurferJeffAtGoogle) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs]: Is there any potential for an SQL injection attack here?

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            ShowQuery(sql, parameters, (processedMb, secs, temperatures));
+
+            // Send spoken response to DialogFlow
+            var top = resultList[0];
+            return DialogflowApp.Tell($"Scanned {processedMb} mega-bytes in {secs:0.0} seconds. " +
+                $"The {temp} temperature in {countryName} in the year {year} was " +
+                $"{(double)top["celcius"]:0.0} degrees celcius, at the {top["name"]} monitoring station.");
+        }
+
+        private static string ExtractAndValidateParameters(ConvRequest req, out string temp, out string year,
```
Comment (SurferJeffAtGoogle) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs]: Needs comment.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/PlatformDescribeHandler.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        /// Handle the intent.
+        /// </summary>
+        /// <param name="req">Conversation request</param>
+        /// <returns></returns>
+        public override async Task<object> Handle(ConvRequest req)
+        {
+            var platform = await Platform.InstanceAsync();
+
+            GetDetailedDescription(platform, out string spokenDescription, out string[] textDescription);
+
```
Comment (SurferJeffAtGoogle) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/PlatformDescribeHandler.cs]: This is more blank lines than I'm used to looking at.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/PlatformDescribeHandler.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            DialogflowApp.Show(string.Join("", textDescription.Select(x => $"<div>{x}</div>")));
+            
+            return DialogflowApp.Tell(spokenDescription);
+        }
+
+        private static void GetDetailedDescription(Platform platform, out string spokenDescription, out string[] textDescription)
```
Comment (SurferJeffAtGoogle) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/PlatformDescribeHandler.cs]: Needs comment.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        /// <summary>
+        /// Initializes class with the conversation.
+        /// </summary>
+        /// <param name="conversation">Conversation</param>
+        public BaseVisionHandler(Conversation conversation) : base(conversation)
+        {
+        }
+
+        // Helper method to combine a list ready for DialogFlow
```
Comment (SurferJeffAtGoogle) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs]: I can't follow all the string concatenation in here.  In the comment, could you show some sample input and output?

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        /// <param name="req"></param>
+        /// <returns></returns>
+        public override async Task<object> Handle(ConvRequest req)
+        {
+            // Unfocus the image
+            _conversation.State.FocusedImage = null;
+
+            // Update state with list of pictures
+            var images = _conversation.State.ImageList;
+            var imagesList = images.Select(x => $"<li><img src=\"{x.Url}\" alt=\"{x.Title}\" style=\"width:200px\" /></li>");
```
Comment (SurferJeffAtGoogle) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: What if the url or title contains a quote character?

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Program.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+        private static void ReadAppSettings()
+        {
+            var config = new ConfigurationBuilder()
+                .SetBasePath(Directory.GetCurrentDirectory())
+                .AddJsonFile("appsettings.json", optional: false)
+                .Build();
+
+            AppSettings = new AppSettings();
+            config.Bind(AppSettings);
```
Comment (SurferJeffAtGoogle) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Program.cs]: What does this do?

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/appsettings.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+﻿{
+  "GoogleCloudSettings": {
+    "ProjectId": "<YOUR PROJECT ID>",
```
Comment (SurferJeffAtGoogle) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/appsettings.json]: By convention, we use hyphenated place holders: YOUR-PROJECT-ID

**Comment by SurferJeffAtGoogle** [comment]
To fix the windows build failure:
1. Clone and build https://github.com/dotnet/codeformatter and put codeformatter.exe in your path.
2. Import BuildTools.psm1.
3. Format-Code

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/ConversationController.cs [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  
+        /// <summary>
+        /// Consructor for Conversation controller.
+        /// </summary>
+        /// <param name="exceptionLogger">Exception logger</param>
+        /// <param name="logger">Regular logger</param>
+        /// <param name="tracer">Tracer</param>
+        public ConversationController(
+            IExceptionLogger exceptionLogger,
+            ILoggerFactory logger,
```
[the remark this answers is not in the document]
Comment (simonz130) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/ConversationController.cs]: When logging, you want to know which component logs the message. If you inject general ILogger to the constructor, it means that either you or DI framework has to create a logger instance and place it in DI container. That logger instance is not attached to a particular component, so all controllers will use the same logger and you won't be able to distinguish easily which controller logged which message. By injecting ILoggerFactory, you can do logger.CreateLogger<T>() (where T is your controller's type) and then use created logger for that controller. This way all the logs from controller X will be distinguishable from controller Y.

[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/ConversationController.cs]: Ok, so we'll keep things the way they are then.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        private ILogger<HomeController> _logger;
+
+        /// <summary>
+        /// Constructor for home controller.
+        /// </summary>
+        /// <param name="exceptionLogger">Exception logger</param>
+        /// <param name="logger">Regular logger</param>
+        public HomeController(
+            IExceptionLogger exceptionLogger,
+            ILoggerFactory logger)
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs]: Ditto

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+            var queue = GetOrCreateClientBufferQueue(clientId);
+
+            var page = string.Empty;
+            try
+            {
+                page = await queue.ReceiveAsync(TimeSpan.FromSeconds(15));
+            }
+            catch(Exception e)
+            {
+                _logger.LogError(e.Message);
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs]: UseGoogleExceptionLogging will catch uncaught exceptions. I wanted to catch an exception and log with HomeController's logger.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            return Content(page, "text/plain");
+        }
+
+        [Route("Test")]
+        public IActionResult Test()
+        {
+            return Content("Hello World!", "text/plain");
+        }
+
+        public static void SetPage(string page)
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs]: Fixed

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+                    }
+                }
+                foreach (var id in toRemove)
+                {
+                    clientBufferQueues.Remove(id);
+                }
+            }
+        }
+
+        private static BufferBlock<string> GetOrCreateClientBufferQueue(string clientId)
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs]: Fixed

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/PlatformDescribeHandler.cs (hunk 4 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/PlatformDescribeHandler.cs]: Fixed

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Program.cs (hunk 7 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Program.cs]: Added a comment explaining what it does

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/appsettings.json (hunk 8 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/appsettings.json]: Fixed

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs (hunk 5 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs]: I added a comment and split out the code a little to make it more obvious

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs (hunk 6 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: These are image urls returned from Google Custom Search, so it won't contain quote character?

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/ConvRequest.cs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace GoogleHomeAspNetCoreDemoServer.Dialogflow
+{
+    public class ConvRequest
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/ConvRequest.cs]: Added some comments

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/ConvRequest.cs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        public static async Task<ConvRequest> ParseAsync(HttpRequest httpRequest)
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/ConvRequest.cs]: Done

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Conversation.cs [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+            }
+            catch (Exception e) when (req.IntentName != "exception.throw")
+            {
+                _exceptionLogger.Log(e);
+                var msg = (e as GoogleApiException)?.Error.Message ?? e.Message;
+                return DialogflowApp.Tell($"Sorry, there's a problem: {msg}");
+            }
+        }
+
+        private BaseHandler FindHandler(string intentName)
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Conversation.cs]: Done

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Conversation.cs [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            var baseHandlerTypes = typeof(BaseHandler).Assembly.GetTypes()
+                .Where(t => t.IsClass && t.IsSubclassOf(typeof(BaseHandler)));
+
+            var typeList = from baseHandlerType in baseHandlerTypes
+                          from attribute in baseHandlerType.GetCustomAttributes(typeof(IntentAttribute), true)
+                          where ((IntentAttribute)attribute).Name == intentName
+                          select baseHandlerType;
+
+            var type = typeList.First();
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Conversation.cs]: Good point, done

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/DialogflowApp.cs [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+                using (_tracer.StartSpan("Conversation"))
+                {
+                    return await conversation.Handle(request);
+                }
+            }
+        }
+
+
+        private Conversation GetOrCreateConversation(ConvRequest convRequest)
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/DialogflowApp.cs]: Done

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/IntentAttribute.cs [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+using System;
+
+namespace GoogleHomeAspNetCoreDemoServer.Dialogflow
+{
+    /// <summary>
+    /// A class level custom attribute that can be used to mark handler classes
+    /// with Dialogflow intent name.
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/IntentAttribute.cs]: Not sure how to rephrase it to be honest. It's basically a class that defines an attribute that you can use to mark handler classes but that's what I say in the comment

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs (hunk 2 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs]: Done

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/PlatformDescribeHandler.cs (hunk 3 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/PlatformDescribeHandler.cs]: Fixed

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/ConversationController.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (SurferJeffAtGoogle) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/ConversationController.cs]: Sorry, my comment wasn't clear.  I'm proposing:
```cs
 public ConversationController(
            IExceptionLogger exceptionLogger,
            ILogger<ConversationController> logger,
            ...)
```

[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/ConversationController.cs]: I see. Ok, that makes sense. Just checked in the fix.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs]: Fixed

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BaseHandler.cs [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+        {
+            _conversation = conversation;
+        }
+
+        /// <summary>
+        /// Handles the conversation request.
+        /// </summary>
+        /// <param name="req"></param>
+        /// <returns></returns>
+        public abstract Task<object> Handle(ConvRequest req);
```
Comment (chrisdunelm) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BaseHandler.cs]: I would strongly suggest changing how this works. There are a few problems currently:
* All handlers are forced to be asynchronous, as they all have to conform to this signature that returns a `Task<...>`.
* The `switch (result)` block in the `Handle` method in the `Conversation` class is pointless as it stands, as all returned values will be `Task<string>`.
* This means some handlers are declared as `async`, but don't use `await`; which is not a good pattern.
* `Task<string>` is the only actual type that must be returned; `Task<object>` is broader than required.
* (And this method should really be called `HandleAsync`).

One way to improve this (in my opinion!) would be to declare two handlers, one sync and one async:
```C#
public virtual Task<string> HandleAsync(ConvRequest req) => return Task.FromResult<string>(null);
public virtual string Handle(ConvRequest req) => return null;
```

Deriving classes only overrides one of these methods, depending on whether they want to do a sync or async implementation.

Then change the code in `Conversation.Handle[Async]` to be more like:
```C#
return handler.Handle(req) ?? await handler.HandleAsync(req) ??
    DialogflowApp.Tell("Error. Handler did not return a valid response.")
```
And the switch statement is no longer required.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace GoogleHomeAspNetCoreDemoServer.Dialogflow.Intents.BigQuery
+{
+    /// <summary>
+    /// Handler for all "bigquery.noaaextreme" DialogFlow intent.
+    /// </summary>
+    [Intent("bigquery.noaaextreme")]
+
```
Comment (chrisdunelm) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs]: nit: remove blank line

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+        /// </summary>
+        /// <param name="col">List to process</param>
+        /// <param name="nonEmptyPrefix">The prefix to add when the list is not empty</param>
+        /// <param name="onEmpty">The string to return when the list is empty.</param>
+        /// <returns></returns>
+        protected static string CombineList(IReadOnlyList<string> col, string nonEmptyPrefix, string onEmpty)
+        {
+            if (col == null || col.Count == 0) return onEmpty;
+ 
+            if (col.Count == 1) return nonEmptyPrefix + " " + col[0];
```
Comment (chrisdunelm) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs]: Use string interpolation:
```C#
if (col.Count == 1) return $"{nonEmptyPrefix} {col[0]}";
```

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+
+            return nonEmptyPrefix + " "  + $"{string.Join(", ", col.Take(col.Count - 1))}, and {col.Last()}";
```
Comment (chrisdunelm) [inline applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs]: Use string interpolation:
```C#
return $"{nonEmptyPrefix} {string.Join(", ", col.Take(col.Count - 1))}, and {col.Last()}";
```

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs (hunk 6 above)
Comment (chrisdunelm) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: [RFC 3986](https://tools.ietf.org/html/rfc3986#section-2.3) specified that quote is not an unreserved character, so they should always be encoded in a URL.
However, the title could contain a quote (or any other) character.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+            return View();
+        }
+
+        [Route("LongPoll")]
+        public async Task<IActionResult> LongPoll()
+        {
+            var clientId = Request.QueryString.Value;
+            var queue = GetOrCreateClientBufferQueue(clientId);
+
+            var page = string.Empty;
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs]: To me, string.Empty makes the intent more clear...

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Conversation.cs [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+        }
+
+        /// <summary>
+        /// Handle a DialogFlow conversation request.
+        /// An instance of this class handles just one conversation,
+        /// which is tracked external to this class, using the DialogFlow sessionID.
+        /// </summary>
+        /// <param name="req">Conversation request</param>
+        /// <returns>A JSON response which is passed back to DialogFlow</returns>
+        public async Task<string> Handle(ConvRequest req)
```
[the remark this answers is not in the document]
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Conversation.cs]: Done

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs (hunk 21 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs]: Good catch. Done.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs (hunk 1 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BigQuery/BigQueryNoaaextremeHandler.cs]: Nothing to change here then.

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs (hunk 6 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: So what do we do here? Do we need to use some encoder to make sure the title is safe?

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (chrisdunelm) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Controllers/HomeController.cs]: Sure, np, leave it as it is then. It certainly isn't incorrect :)

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs (hunk 6 above)
Comment (SurferJeffAtGoogle) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: I looked at using a razor page, but trying to use razor outside of MVC is a total hassle.

How about using https://docs.microsoft.com/en-us/dotnet/api/system.net.webutility.htmlencode?view=netcore-2.0#System_Net_WebUtility_HtmlEncode_System_String_?

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BaseHandler.cs (hunk 20 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/BaseHandler.cs]: Although I wasn't happy about having two methods for essentially the same thing, I think there's no better alternative. So, I went ahead and implemented what you suggested. Thanks!

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs (hunk 22 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs]: Fixed

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs (hunk 23 above)
Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/BaseVisionHandler.cs]: Fixed

File: applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs (hunk 6 above)
Comment (SurferJeffAtGoogle) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: This is the final remaining thing I'd like to see fixed before merging.

Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: Yes, me too :-) I plan to fix the rest of the comments today, so we can merge it later today or tomorrow.

Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: Fixed. All the titles are encoded now.

Comment (meteatamel) [reply applications/googlehome-meets-dotnetcontainers/GoogleHomeAspNetCoreDemoServer/Dialogflow/Intents/Vision/VisionBacktoallHandler.cs]: All the titles are encoded now
